src/js/angular/core/directives/languageselector/language-selector.directive.js

```diff
@@ -31,12 +33,44 @@ function languageSelector($translate, LocalStorageAdapter, LSKeys, $languageServ
             // translated they will be translated in English
             $translate.fallbackLanguage('en');
 
+            function requiresReload(tab) {
+                return tab.drawnPlugin === 'charts' || tab.drawnPlugin === 'pivot-table-plugin';
+            }
+
             $scope.changeLanguage = function (lang) {
-                $scope.selectedLang = lang
-                $translate.use(lang.key);
-                LocalStorageAdapter.set(LSKeys.PREFERRED_LANG, lang.key);
-                $languageService.setLanguage($scope.selectedLang.key);
-                $scope.$broadcast('language-changed', {locale: lang.key});
+                const yasgui = YasguiComponentDirectiveUtil.getOntotextYasguiElement(QUERY_EDITOR_ID);
+                const promise = yasgui ? yasgui.yasguiComponent.getActivePluginTab() : Promise.resolve(null);
+
+                promise.then((tab) => {
+                    const shouldReload = tab ? requiresReload(tab) : false;
+                    // page reload required on some plugin tabs
+                    if (yasgui && shouldReload && $languageService.getLanguage() !== lang.key) {
+                        ModalService.openSimpleModal({
+                            title: $translate.instant('query.editor.language.change.warning.title'),
+                            message: $translate.instant('query.editor.reload.page.warning'),
+                            warning: true
+                        }).result.then(function () {
+                            $scope.selectedLang = lang;
+                            $translate.use(lang.key);
+                            LocalStorageAdapter.set(LSKeys.PREFERRED_LANG, lang.key);
+                            $languageService.setLanguage($scope.selectedLang.key);
+                            location.reload();
+                        });
+                    } else if ($languageService.getLanguage() !== lang.key) {
+                        $scope.selectedLang = lang;
+                        $translate.use(lang.key);
+                        LocalStorageAdapter.set(LSKeys.PREFERRED_LANG, lang.key);
+                        $languageService.setLanguage($scope.selectedLang.key);
+                        $scope.$broadcast('language-changed', {locale: lang.key});
+                    }
+                });
+
+                if (!yasgui && $languageService.getLanguage() !== lang.key) {
+                    $scope.selectedLang = lang;
+                    $translate.use(lang.key);
+                    LocalStorageAdapter.set(LSKeys.PREFERRED_LANG, lang.key);
+                    $languageService.setLanguage($scope.selectedLang.key);
```

Comment:
Is there a reason for repeating these 4 lines three times? Think of a way to improve this!